### PR TITLE
Customizable template and remove Google Fonts

### PR DESCRIPTION
**Pull request by markuspoerschke**
This Merge Request contains the following changes:
* Remove usage of Google Fonts (can be reenabled if needed)
* Template can be now customized. New blocks were added to the template.

To override the template create a new twig file:

```twig
{# templates/bundles/NelmioApiDocBundle/SwaggerUI/index.html.twig #}

{% extends '@!NelmioApiDoc/SwaggerUi/index.html.twig' %}

{% block stylesheets %}
    {{ parent() }}
    <link rel="stylesheet" href="{{ asset('css/custom-swagger-styles.css') }}">
{% endblock stylesheets %}

{% block javascripts %}
    {{ parent() }}
    <script type="text/javascript" src="{{ asset('js/custom-request-signer.js') }}"></script>
{% endblock javascripts %}
```

### CODE REVIEW

File: Resources/views/SwaggerUi/index.html.twig [hunk 1]
```diff
@@ -8,54 +8,63 @@ file that was distributed with this source code. #}
...
 <html>
 <head>
-    <meta charset="UTF-8">
-    <title>{{ swagger_data.spec.info.title }} - NelmioApiDocBundle</title>
+    {% block meta %}
+        <meta charset="UTF-8">
+    {% endblock meta %}
+    <title>{% block title %}{{ swagger_data.spec.info.title }}{% endblock title %}</title>
 
-    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,700|Source+Code+Pro:300,600|Titillium+Web:400,600,700">
```
Comment (dbu) [inline Resources/views/SwaggerUi/index.html.twig]: did you remove this on purpose?

File: Resources/views/SwaggerUi/index.html.twig [hunk 2]
```diff
@@ -8,54 +8,63 @@ file that was distributed with this source code. #}
...
-        &copy; 2017 <a href="https://api-platform.com">Api-Platform</a>
-    </div>
-
-    <script src="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui-bundle.js') }}"></script>
-    <script src="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui-standalone-preset.js') }}"></script>
-    <script src="{{ asset('bundles/nelmioapidoc/init-swagger-ui.js') }}"></script>
+    {% block javascripts %}
+        <script src="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui-bundle.js') }}"></script>
+        <script src="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui-standalone-preset.js') }}"></script>
+        <script src="{{ asset('bundles/nelmioapidoc/init-swagger-ui.js') }}"></script>
```
Comment (dbu) [inline Resources/views/SwaggerUi/index.html.twig]: am i right assuming that the init-swagger-ui js is a thing that needs to be customized quite often? would it make sense to split this into a sub-block so that it can be customized without touching the rest of the scripts block? although, having to copy the two lines above is not a big issue either.

File: Controller/SwaggerUiController.php [hunk 3]
```diff
@@ -24,10 +24,12 @@
 
     private $twig;
 
+    private $twigTemplate;
+
     /**
      * @param ContainerInterface $generatorLocator
      */
-    public function __construct($generatorLocator, \Twig_Environment $twig)
+    public function __construct($generatorLocator, \Twig_Environment $twig, string $twigTemplate)
```
Comment (dbu) [inline Controller/SwaggerUiController.php]: there is one failing test about this being called with only 2 arguments.

its an edge case for BC considerations, but maybe it would make sense to set the default value on this constructor argument as well?

File: Resources/views/SwaggerUi/index.html.twig [hunk 4]
```diff
@@ -8,54 +8,70 @@ file that was distributed with this source code. #}
...
+    {% block meta %}
+        <meta charset="UTF-8">
+    {% endblock meta %}
+    <title>{% block title %}{{ swagger_data.spec.info.title }}{% endblock title %}</title>
 
-    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,700|Source+Code+Pro:300,600|Titillium+Web:400,600,700">
-    <link rel="stylesheet" href="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui.css') }}">
-    <link rel="stylesheet" href="{{ asset('bundles/nelmioapidoc/style.css') }}">
+    {% block fonts %}
+        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,700|Source+Code+Pro:300,600|Titillium+Web:400,600,700">
```
[the remark this answers is not in the document]
Comment (dbu) [reply Resources/views/SwaggerUi/index.html.twig]: i asked about it, because i did not see the other thing. with the change in swagger proper, i agree to remove it. and add an example in the customization documentation how to overwrite the stylesheets block to add the font and then call the parent block for the other stylesheets.

File: Resources/views/SwaggerUi/index.html.twig [hunk 5]
```diff
@@ -8,54 +8,70 @@ file that was distributed with this source code. #}
...
+    {% block javascripts %}
+        <script src="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui-bundle.js') }}"></script>
+        <script src="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui-standalone-preset.js') }}"></script>
+    {% endblock javascripts %}
 
-    <script src="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui-bundle.js') }}"></script>
-    <script src="{{ asset('bundles/nelmioapidoc/swagger-ui/swagger-ui-standalone-preset.js') }}"></script>
-    <script src="{{ asset('bundles/nelmioapidoc/init-swagger-ui.js') }}"></script>
+    {% block swagger_initialization %}
+        <script src="{{ asset('bundles/nelmioapidoc/init-swagger-ui.js') }}"></script>
```
[the remark this answers is not in the document]
Comment (dbu) [reply Resources/views/SwaggerUi/index.html.twig]: it used to not be inlined though. why do you want to inline it now?

File: Controller/SwaggerUiController.php [hunk 6]
```diff
@@ -20,14 +20,20 @@
...
 
+    private $twigTemplate;
+
     /**
      * @param ContainerInterface $generatorLocator
+     * @param \Twig_Environment  $twig
+     * @param string             $twigTemplate
      */
-    public function __construct($generatorLocator, \Twig_Environment $twig)
+    public function __construct($generatorLocator, \Twig_Environment $twig, string $twigTemplate = self::TWIG_TEMPLATE_DEFAULT)
```
[the remark this answers is not in the document]
Comment (dbu) [reply Controller/SwaggerUiController.php]: oh cool, then we don't need to introduce configuration and this PR can be reduced to documentation (including a pointer to the symfony doc), plus the twig blocks to allow the extending template to overwrite things .

File: Resources/views/SwaggerUi/index.html.twig (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbu) [reply Resources/views/SwaggerUi/index.html.twig]: resp. adjust the existing example to tell that thing

**Comment by dbu** [comment]
can you please add 2 entries to the changelog? one for that the template can now be overwritten and extended, and one for the removal of the google font inclusion?

File: Resources/views/SwaggerUi/index.html.twig (hunk 4 above)
[the remark this answers is not in the document]
Comment (GuilhemN) [reply Resources/views/SwaggerUi/index.html.twig]: Thinking about it again, I think we have more important to document than how to change a font, imo we should just merge

File: Resources/views/SwaggerUi/index.html.twig (hunk 5 above)
[the remark this answers is not in the document]
Comment (GuilhemN) [reply Resources/views/SwaggerUi/index.html.twig]: It would allow us to remove the need for `swagger-data` (could be inlined directly in the script). Also as pointed by @magnetik, we could this way have a block to change constructor parameters (I don't think this is possible after the instanciation)

File: Resources/views/SwaggerUi/index.html.twig [hunk 7]
```diff
@@ -8,54 +8,63 @@ file that was distributed with this source code. #}
...
+    {% endblock svg_icons %}
     <header>
-        <a id="logo" href="https://github.com/nelmio/NelmioApiDocBundle"><img src="{{ asset('bundles/nelmioapidoc/logo.png') }}" alt="NelmioApiDocBundle"></a>
+        {% block header %}
+            <a id="logo" href="https://github.com/nelmio/NelmioApiDocBundle"><img src="{{ asset('bundles/nelmioapidoc/logo.png') }}" alt="NelmioApiDocBundle"></a>
+        {% endblock header %}
     </header>
 
-    <div id="swagger-ui" class="api-platform"></div>
+    {% block swagger_ui %}
```
[the remark this answers is not in the document]
Comment (markuspoerschke) [reply Resources/views/SwaggerUi/index.html.twig]: I think the actual use case is to add some content below or above.

I had this case where I needed to add a very special login form.

This can be done like this:

```twig
{% block swagger_ui %}
    {{ parent() }}
    <div class="some-class">
        <!-- Some additional HTML elements -->
   </div>
{% endblock %}}
```

File: Controller/SwaggerUiController.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (markuspoerschke) [reply Controller/SwaggerUiController.php]: I did not know about that feature! Thanks a lot. I will change my PR accordingly!

File: Resources/views/SwaggerUi/index.html.twig (hunk 4 above)
[the remark this answers is not in the document]
Comment (markuspoerschke) [reply Resources/views/SwaggerUi/index.html.twig]: I am now uncertain how to treat this. In my opinion there are two options:

1. Keep it as it is in this PR: fonts block and include google fonts + add documentation how to remove the font
2. Remove the font and document (in CHANGELOG) how to readd it using the template override.

I would prefer option 2.

@GuilhemN @dbu What should I do?

[the remark this answers is not in the document]
Comment (GuilhemN) [reply Resources/views/SwaggerUi/index.html.twig]: I prefer being as simple as possible so I would go for op. 2 :+1:

[the remark this answers is not in the document]
Comment (dbu) [reply Resources/views/SwaggerUi/index.html.twig]: i also prefer version 2, remove it and add a changelog entry about it (plus one about the template blocks and overwriting)

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,6 +1,55 @@
...
+            .swagger-ui .scopes h2,
+            .swagger-ui .errors-wrapper hgroup h4 {
+                font-family: Open Sans,sans-serif!important;
+            }
+        </style>
+    {% endblock stylesheets %}
+    ```
+
+* The Twig template for the Swagger UI now contains blocks to make it easier to overwrite certain parts.
+  See the official documentation how to do this.
```
Comment (dbu) [inline CHANGELOG.md]: can we make this link to the relevant documentation page? (markdown link format is `[official documentation](https://symfony.com/...)`)

File: DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -11,6 +11,7 @@
 
 namespace Nelmio\ApiDocBundle\DependencyInjection;
 
+use Nelmio\ApiDocBundle\Controller\SwaggerUiController;
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: this is now unused

File: DependencyInjection/NelmioApiDocExtension.php [hunk 10]
```diff
@@ -105,6 +105,8 @@ public function load(array $configs, ContainerBuilder $container)
                 ->addTag(sprintf('nelmio_api_doc.describer.%s', $area), ['priority' => -200]);
         }
 
+        $container->setParameter('nelmio_api_doc.template', $config['config']['template']);
```
Comment (dbu) [inline DependencyInjection/NelmioApiDocExtension.php]: this no longer exists and should be removed

File: Resources/config/services.xml [hunk 11]
```diff
@@ -8,6 +8,7 @@
         <service id="nelmio_api_doc.controller.swagger_ui" class="Nelmio\ApiDocBundle\Controller\SwaggerUiController" public="true">
             <argument type="service" id="nelmio_api_doc.generator_locator" />
             <argument type="service" id="twig" />
+            <argument>%nelmio_api_doc.template%</argument>
```
Comment (dbu) [inline Resources/config/services.xml]: this no longer exists and should be removed

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,6 +1,55 @@
...
 
+3.3.0 (unreleased)
+------------------
+
+* Usage of Google Fonts was removed. System fonts `serif` / `sans` will be used instead. 
+  This can lead to a different look on different operating systems.
+  You can add the Google Fonts again manually by overriding the template:
+  
+  * Create a new file ``
+  * Add the following content:
```
Comment (dbu) [inline CHANGELOG.md]: this is very large for a changelog entry. can you add a FAQ entry instead: "How to change the font" and link to that instead?

File: Resources/doc/customization.rst [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+The following example will add additional scripts and a custom style to the template.
+Just create a file ``templates/bundles/NelmioApiDocBundle/views/SwaggerUI/index.html.twig``.
+
+.. code-block:: twig
+
+    {# templates/bundles/NelmioApiDocBundle/views/SwaggerUI/index.html.twig #}
+
+    {#
+        To avoid a "reached nested level" error an exclamation mark `!` has to be added
```
Comment (dbu) [inline Resources/doc/customization.rst]: great way of explaining this, i like it.

File: Resources/views/SwaggerUi/index.html.twig (hunk 7 above)
[the remark this answers is not in the document]
Comment (GuilhemN) [reply Resources/views/SwaggerUi/index.html.twig]: Ok :+1:

**Comment by GuilhemN** [comment]
This is really great and an awaited feature, thank you!

**Comment by numediaweb** [comment]
the `template` directory is for Symfony 4: any BC for Symfony 3.4?

**Comment by GuilhemN** [comment]
@numediaweb our example is using the flex directory structure, for 3.4, just put the file in `app/Resources/NelmioApiDocBundle/views/SwaggerUI/index.html.twig` instead.

**Comment by bobvandevijver** [comment]
@GuilhemN I really want to use this, but it has not yet been tagged... Can you create a release? 😄 (I really do not want to add "unstable" dependencies)

**Comment by GuilhemN** [comment]
I'd like to merge https://github.com/nelmio/NelmioApiDocBundle/pull/1377 first, but then sure, I'll release 3.4 😄
